[spec/function.dd] Add scope parameter section

ntrel opened the pull request:
* Tweak existing wording
* Mention `@safe` & -dip1000 needed for enforcement

*Update:*
* Split existing `return ref` example.

jacob-carlborg commented on spec/function.dd:
```diff
@@ -1238,9 +1238,13 @@ int foo(in int x, out int y, ref int z, int q);
+    Enforced only for `@safe` functions.
+    Ignored for parameters with no references.
+    Note: The `scope` attribute is currently only enforced by `dmd`
+    when the `-dip1000` switch is passed.
```
Doesn't `scope` on delegate has some affect even without `-dip1000`?

thewilsonator replied:
Delegates are reference types though. But the elision of GC allocation of the closure should be mentioned.

ntrel replied:
@jacob-carlborg I've now documented this, in a more general way. Walter has said that heap allocation for other reference types besides delegates can be elided too, e.g. for array concatenation:
https://forum.dlang.org/post/ofo9f5$2bt7$1@digitalmars.com

ntrel commented:
@thewilsonator Sorry I forgot about this. Rebased! (At least I get CircleCI notifications now when my builds fail :-) ).

MoonlightSentinel commented on spec/function.dd:
```diff
@@ -1242,9 +1242,18 @@ int foo(in int x, out int y, ref int z, int q);
+    As the parameter does not escape, the compiler can avoid heap-allocation of a unique argument
+    to a `scope` parameter. Due to this, such a function call can become `@nogc`.
+    This is typically done when passing a delegate literal to a `scope` parameter.
+    $(B Note): Escape analysis is currently only done by `dmd`
+    when the `-dip1000` switch is passed.
```
This statement isn't exactly true, there is some scope analysis without `-dip1000` (esp. given that [`-dip25` is now partially enabled by default](https://dlang.org/changelog/2.092.0.html#dip25)). DIP1000 enables some idioms that are otherwise forbidden, e.g. pointers to local variables.

ntrel replied:
-dip25 only analyses escaping refs, not scope parameters. But I changed it to 'scope escape analysis...' for clarity.

MoonlightSentinel commented on spec/function.dd:
```diff
@@ -1408,7 +1417,7 @@ pure void f()
 $(H3 $(LNAME2 return-ref-parameters, Return Ref Parameters))
 
-        $(P Note: The `return` attribute is currently only enforced by `dmd`
+        $(P $(B Note): The `return` attribute is currently only enforced by `dmd`
```
This could mention the deprecations issued since v2.092 ([which partially enables DIP25 by default](https://dlang.org/changelog/2.092.0.html#dip25)

ntrel replied:
Thanks. I decided to just remove the note, as the deprecation warnings will become errors at some point.

ntrel commented on spec/function.dd:
```diff
@@ -1242,9 +1242,18 @@ int foo(in int x, out int y, ref int z, int q);
+    Escape analysis is done for `@safe` functions, for other functions `scope` semantics must
+    be manually enforced.
+    As the parameter does not escape, the compiler can avoid heap-allocation of a unique argument
+    to a `scope` parameter. Due to this, such a function call can become `@nogc`.
```
Yes, please see 5843070 for the changed wording.

MoonlightSentinel replied:
That's definitly better but I would change the example because passing `new T` to a scope parameter isn't `@nogc` (yet), You could also mention array literals (which were implemented recently)

ntrel replied:
OK, I've now listed several things including array literals that can potentially avoid allocation.

ntrel commented:
The scope parameter changes were getting a bit big for the function parameter table, so I have moved them to their own section and split out an example from the `return scope` docs.
